File: src/main/java/io/github/chriszhong/twitter/AppConfig.java
package io.github.chriszhong.twitter;

import javax.servlet.http.HttpSession;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Scope;
import org.springframework.context.annotation.ScopedProxyMode;
import org.springframework.social.config.annotation.EnableSocial;
import org.springframework.social.connect.Connection;
import org.springframework.social.connect.ConnectionRepository;
import org.springframework.social.connect.UsersConnectionRepository;
import org.springframework.social.twitter.api.Twitter;
import org.springframework.web.context.WebApplicationContext;

@Configuration
@EnableSocial
public class AppConfig {

	private static final Logger logger = LoggerFactory.getLogger(AppConfig.class);

	@Bean
	@Scope(scopeName = WebApplicationContext.SCOPE_REQUEST, proxyMode = ScopedProxyMode.INTERFACES)
	public ConnectionRepository connectionRepository(UsersConnectionRepository usersConnectionRepository, HttpSession httpSession) {
		ConnectionRepository connectionRepository = usersConnectionRepository.createConnectionRepository(httpSession.getId());
		logger.debug("Using `{}` for '{}'", connectionRepository, httpSession.getId());
		return connectionRepository;
	}

	@Bean
	@Scope(scopeName = WebApplicationContext.SCOPE_REQUEST, proxyMode = ScopedProxyMode.INTERFACES)
	public Twitter twitter(ConnectionRepository repository) {
		Connection<Twitter> connection = repository.findPrimaryConnection(Twitter.class);
		return connection != null ? connection.getApi() : null;
	}

}

File: src/main/java/io/github/chriszhong/twitter/AppConfig.kt
package io.github.chriszhong.twitter

import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Scope
import org.springframework.context.annotation.ScopedProxyMode
import org.springframework.social.config.annotation.EnableSocial
import org.springframework.social.connect.ConnectionRepository
import org.springframework.social.connect.UsersConnectionRepository
import org.springframework.social.twitter.api.Twitter
import org.springframework.web.context.WebApplicationContext
import javax.servlet.http.HttpSession

@Configuration
@EnableSocial
open class AppConfig {

    @Bean
    @Scope(scopeName = WebApplicationContext.SCOPE_REQUEST, proxyMode = ScopedProxyMode.INTERFACES)
    open fun connectionRepository(usersConnectionRepository: UsersConnectionRepository, httpSession: HttpSession): ConnectionRepository {
        val connectionRepository = usersConnectionRepository.createConnectionRepository(httpSession.id)
        logger.debug("Using `{}` for '{}'", connectionRepository, httpSession.id)
        return connectionRepository
    }

    @Bean
    @Scope(scopeName = WebApplicationContext.SCOPE_REQUEST, proxyMode = ScopedProxyMode.INTERFACES)
    open fun twitter(repository: ConnectionRepository): Twitter? {
        val connection = repository.findPrimaryConnection(Twitter::class.java)
        return connection?.api
    }

    companion object {
        private val logger = LoggerFactory.getLogger(AppConfig::class.java)
    }
}
